File: app/backtesting/strategies/ema_backtest.py
import pandas as pd


class EMABacktestStrategy:

    def __init__(
        self,
        fast_period=5,
        slow_period=10
    ):

        self.fast_period = fast_period
        self.slow_period = slow_period

    def generate_signals(
        self,
        dataframe: pd.DataFrame
    ):

        dataframe["fast_ema"] = (
            dataframe["close"]
            .ewm(span=self.fast_period)
            .mean()
        )

        dataframe["slow_ema"] = (
            dataframe["close"]
            .ewm(span=self.slow_period)
            .mean()
        )

        dataframe["signal"] = 0

        dataframe.loc[
            dataframe["fast_ema"]
            > dataframe["slow_ema"],
            "signal"
        ] = 1

        dataframe.loc[
            dataframe["fast_ema"]
            < dataframe["slow_ema"],
            "signal"
        ] = -1

        return dataframe

File: backend/src/backtesting/strategies/emaBacktest.ts
import { CSVRow } from "../data/csvLoader";
import { calculateEMA } from "../../utils/indicators";

export class EMABacktestStrategy {
  private fastPeriod: number;
  private slowPeriod: number;

  constructor(fastPeriod: number = 5, slowPeriod: number = 10) {
    this.fastPeriod = fastPeriod;
    this.slowPeriod = slowPeriod;
  }

  generateSignals(dataframe: CSVRow[]): CSVRow[] {
    const closes = dataframe.map((row) => row.close);
    const fastEma = calculateEMA(closes, this.fastPeriod);
    const slowEma = calculateEMA(closes, this.slowPeriod);

    for (let i = 0; i < dataframe.length; i++) {
      const row = dataframe[i];
      row.fast_ema = fastEma[i];
      row.slow_ema = slowEma[i];

      if (row.fast_ema > row.slow_ema) {
        row.signal = 1;
      } else if (row.fast_ema < row.slow_ema) {
        row.signal = -1;
      } else {
        row.signal = 0;
      }
    }

    return dataframe;
  }
}
export default EMABacktestStrategy;
